domain/model/service/service_test.go

```diff
@@ -1112,3 +1122,93 @@ func (s *serviceSuite) TestDefaultModelCloudNameAndCredential(c *gc.C) {
 		Name:  "test-cred",
 	})
 }
+
+// TestWatchActivatedModels verifies that WatchActivatedModels correctly sets up a watcher
+// that emits events for activated models when the watcher receives change events.
+func (s *serviceSuite) TestWatchActivatedModels(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	ctx := context.Background()
+	svc := NewWatchableService(
+		s.mockState,
+		s.mockModelDeleter,
+		loggertesting.WrapCheckLog(c),
+		s.mockWatcherFactory,
+	)
+
+	// Set up necessary mock return values.
+	s.mockState.EXPECT().GetAllActivatedModelsUUIDQuery().Return(
+		"SELECT uuid from model WHERE activated = true",
+	)
+
+	changes := make(chan []string, 1)
+	activatedModelUUID1 := modeltesting.GenModelUUID(c)
+	activatedModelUUID2 := modeltesting.GenModelUUID(c)
+	activatedModelUUIDs := []coremodel.UUID{activatedModelUUID1, activatedModelUUID2}
+	activatedModelUUIDsStr := transform.Slice(activatedModelUUIDs, func(uuid coremodel.UUID) string {
+		return uuid.String()
+	})
+	changes <- activatedModelUUIDsStr
+	s.mockStringsWatcher.EXPECT().Changes().AnyTimes().Return(changes)
+
+	s.mockWatcherFactory.EXPECT().NewNamespaceMapperWatcher("model", changestream.Changed,
+		gomock.Any(), gomock.Any()).Return(s.mockStringsWatcher, nil)
+
+	// Verifies that the service returns a watcher with the correct model UUIDs string.
+	watcher, err := svc.WatchActivatedModels(ctx)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(<-watcher.Changes(), gc.DeepEquals, activatedModelUUIDsStr)
+}
+
+func (s *serviceSuite) createMockChangeEventsFromUUIDs(uuids ...coremodel.UUID) []changestream.ChangeEvent {
+	events := make([]changestream.ChangeEvent, len(uuids))
+	for i, uuid := range uuids {
+		event := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
```

Comment:
This func should take the mock controller as an argument and not pull it off other things. The way this is assumes a lot about how the caller has set things up.

Reply:
Good point, thanks tom!